lib/CGameState.cpp

```diff
@@ -1608,6 +1610,64 @@ void CGameState::giveCampaignBonusToHero(CGHeroInstance * hero)
 	}
 }
 
+void CGameState::removeSightnObj(const CGObjectInstance * obj)
+{
+	addSightObj(obj, false);
+}
+
+void CGameState::addSightObj(const CGObjectInstance * obj, bool add)
+{
+	if(!vstd::contains(players, obj->tempOwner))
+		return;
+
+	auto p = getPlayer(obj->tempOwner);
+	if(p->status != EPlayerStatus::INGAME)
+		return;
+
+	addSightObj(p->team, obj, add);
+}
+
+void CGameState::addSightObj(TeamID team, const CGObjectInstance * obj, bool add)
+{
+	logGlobal->debug("%s object on sight map: %d", add ? "Addding" : "Removing", obj->id);
+	auto ts = getTeam(team);
+	std::unordered_set<int3, ShashInt3> tiles;
+	getTilesInRange(tiles, obj->getSightCenter(), obj->getSightRadius());
+	for(int3 t : tiles)
+	{
+		auto old = ts->fogOfWarMap[t.x][t.y][t.z];
+		if(add)
+		{
+			if(ts->fogOfWarMap[t.x][t.y][t.z] == 0)
+				ts->fogOfWarMap[t.x][t.y][t.z] = 2;
```

Comment:
0 mean that tile in hidden from team.
1 mean that it's visible.
Anything more than 1 mean there is N objects with sight over that tile so it's can't be covered by darkness